### PR TITLE
Faster GPU frozen BatchNorm

### PR DESCRIPTION
**Pull request by ptrendx**
## Description ##
This PR introduces an improved implementation of GPU BatchNorm when `use_global_stats` is `True`

Performance results (using V100 PCIe card, shape of data = `(208, 64, 112, 112)`)

dtype = `float32`

| FWD/BWD | axis | Old implementation time (ms) | New implementation time (ms) | speedup |
| - | - | - | - | - |
| FWD | 0 | 3.67 | 2.31 | 1.59x |
| BWD | 0 | 6.60 | 2.28 | 2.9x |
| FWD | 1 | 4.92 | 2.21 | 2.23x |
| BWD | 1 | 8.61 | 2.23 | 3.86x |
| FWD | 2 | 11.52 | 2.21 | 5.21x |
| BWD | 2 | 21.36 | 2.74 | 7.8x |
| FWD | 3 | 1010.55 | 2.24 | 451x |
| BWD | 3 | 1901.58 | 2.77 | 686.5x |

dtype = `float16`

| FWD/BWD | axis | Old implementation time (ms) | New implementation time (ms) | speedup |
| - | - | - | - | - |
| FWD | 0 | 3.1 | 0.92 | 3.37x |
| BWD | 0 | 6.56 | 1.04 | 6.31x |
| FWD | 1 | 4.34 | 0.92 | 4.72x |
| BWD | 1 | 9.21 | 1.34 | 6.87x |
| FWD | 2 | 10.23 | 0.92 | 11.12x |
| BWD | 2 | 22.61 | 1.60 | 14.13x |
| FWD | 3 | 948.1 | 1.01 | 939x |
| BWD | 3 | 2061.9 | 1.32 | 1562x |

@Kh4L @Jerryzcn FYI

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by Jerryzcn** [comment]
@haojin2 @sxjscience

**Comment by Jerryzcn** [comment]
Have you tried training MaskRCNN with this?

**Comment by ptrendx** [comment]
@Jerryzcn Not yet.

File: src/operator/nn/batch_norm.cu [hunk 1]
```diff
@@ -44,9 +46,30 @@
...
+}
+
+template <>
+MSHADOW_XINLINE double variance_to_invstd(double var, double eps) {
+  return rsqrt(var + eps);
+}
+
+template <typename AccReal>
+MSHADOW_XINLINE AccReal invstd_to_variance(AccReal invstd, AccReal eps) {
+  return 1.0f / (invstd * invstd) - eps;
```
Comment (Kh4L) [inline src/operator/nn/batch_norm.cu]: What about:
```suggestion
  return static_cast<AccReal>(1.0) / (invstd * invstd) - eps;
```

Comment (ptrendx) [reply src/operator/nn/batch_norm.cu]: Good suggestion.

File: src/common/cuda_utils.h [hunk 2]
```diff
@@ -764,27 +764,86 @@ __device__ inline DType ldg(const DType* address) {
...
+ *              false, then only thread 0 has the valid result. Defaults to true.
+ * \param value - value from each thread to be reduced
+ * \param redfun - function used to perform reduction
+ */
+template <int NTHREADS, bool all_reduce = true, typename OP, typename T>
+__device__ inline T reduce(const T& value, OP redfun) {
+  static_assert(NTHREADS <= warp_size * warp_size,
+                "Number of threads too large for reduction");
+  __shared__ T scratch[NTHREADS / warp_size];
+  const int my_id = threadIdx.x % warp_size;
```
Comment (apeforest) [inline src/common/cuda_utils.h]: Could we use a more informative names other than my_*?

**Comment by Jerryzcn** [comment]
Could we get some convergence result on fasterrcnn before merging thanks!

**Comment by Kh4L** [comment]
@Jerryzcn it's in progress, I'll post results here

**Comment by ptrendx** [comment]
@Kh4L Could you post results of your runs?

**Comment by Kh4L** [comment]
@ptrendx @Jerryzcn sorry for the delay, please find it here:

[joblog.log](https://github.com/apache/incubator-mxnet/files/4195098/joblog.log)

**Comment by Jerryzcn** [comment]
Lgtm

**Comment by Jerryzcn** [comment]
BTW, were you able to find the bug that cause master MXNet to fail on mask rcnn?

**Comment by Kh4L** [comment]
@Jerryzcn no, we didn't really look into it, as we thought that you were taking over this task

There is an open issue here https://github.com/apache/incubator-mxnet/issues/17485

**Comment by ptrendx** [comment]
Reviving this PR. @apeforest I did change `my_id` to `thread_idx_in_warp`, but other cases of `my_*` variables are actually quite intuitive (and basically mean a private value of a given thing owned by a thread).

**Comment by ptrendx** [comment]
@mxnet-bot run ci [centos-cpu]

**Comment by ptrendx** [comment]
@mxnet-bot run ci [centos-cpu, unix-gpu]

**Comment by ptrendx** [comment]
@mxnet-bot run ci [unix-gpu]

**Comment by ptrendx** [comment]
@apeforest Is this PR good to go now?

**Comment by ptrendx** [comment]
@apeforest Gentle ping

**Comment by szha** [comment]
hmm not sure what's happening with the CI.

**Comment by ptrendx** [comment]
@mxnet-bot run ci [centos-cpu, miscellaneous]

**Comment by ptrendx** [comment]
@mxnet-bot run ci [centos-cpu, miscellaneous]
